includes/class-reader-activation.php

```diff
@@ -855,5 +857,133 @@ public static function get_client_id() {
 		// phpcs:ignore WordPressVIPMinimum.Variables.RestrictedVariables.cache_constraints___COOKIE
 		return isset( $_COOKIE[ NEWSPACK_CLIENT_ID_COOKIE_NAME ] ) ? sanitize_text_field( $_COOKIE[ NEWSPACK_CLIENT_ID_COOKIE_NAME ] ) : false;
 	}
+
+	/**
+	 * Update content metadata after a contact's lists are updated.
+	 *
+	 * @param string        $provider        The provider name.
+	 * @param string        $email           Contact email address.
+	 * @param string[]      $lists_to_add    Array of list IDs to subscribe the contact to.
+	 * @param string[]      $lists_to_remove Array of list IDs to remove the contact from.
+	 * @param bool|WP_Error $result          True if the contact was updated or error if failed.
+	 */
+	public static function newspack_newsletters_update_contact_lists( $provider, $email, $lists_to_add, $lists_to_remove, $result ) {
+		switch ( $provider ) {
+			case 'active_campaign':
+				if ( true === $result && method_exists( '\Newspack_Newsletters_Subscription', 'add_contact' ) && method_exists( '\Newspack_Newsletters_Subscription', 'get_contact_lists' ) ) {
+					$current_lists = \Newspack_Newsletters_Subscription::get_contact_lists( $email );
+					// The add_contact method is idempotent, effectively being an upsertion.
+					\Newspack_Newsletters_Subscription::add_contact( [ 'email' => $email ], $current_lists );
+				}
+				break;
+		}
+	}
+
+	/**
+	 * Modify metadata for newsletter contact creation.
+	 *
+	 * @param string   $provider The provider name.
+	 * @param array    $contact  {
+	 *    Contact information.
+	 *
+	 *    @type string   $email    Contact email address.
+	 *    @type string   $name     Contact name. Optional.
+	 *    @type string[] $metadata Contact additional metadata. Optional.
+	 * }
+	 * @param string[] $selected_list_ids    Array of list IDs to subscribe the contact to.
+	 */
+	public static function newspack_newsletters_contact_data( $provider, $contact, $selected_list_ids ) {
+		switch ( $provider ) {
+			case 'active_campaign':
+				$metadata = [];
+				if ( is_user_logged_in() ) {
+					$metadata['NP_Account'] = get_current_user_id();
+				}
+
+				// If it's a new contact, add a registration or signup date.
+				$is_new_contact = null;
+				try {
+					if ( method_exists( '\Newspack_Newsletters_Subscription', 'existing_contact_data' ) ) {
+						$existing_contact = \Newspack_Newsletters_Subscription::existing_contact_data( $contact['email'] );
+						if ( is_wp_error( $existing_contact ) ) {
+							Logger::log( 'Adding metadata to a new contact.' );
+							$is_new_contact = true;
+							if ( empty( $selected_list_ids ) ) {
+								// Registration only, as a side effect of Reader Activation.
+								$contact['metadata']['NP_Registration Date'] = gmdate( 'm/d/Y' );
```

Comment:
Which standard are we following for naming the metadata keys? Shouldn't it be either spaces or underscores?

Reply:
This prefix was suggested by a partner, let's keep it for now.